### PR TITLE
Postgres test migration fix

### PR DESCRIPTION
**Pull request by aniketfuryrocks**

### CODE REVIEW

File: bench/src/metrics.rs [hunk 1]
```diff
@@ -35,6 +35,9 @@ impl AddAssign<&Self> for Metric {
 
 impl DivAssign<u64> for Metric {
     fn div_assign(&mut self, rhs: u64) {
+        if rhs == 0 {
+            return;
+        }
```
Comment (mschneider) [inline bench/src/metrics.rs:40]: where is this used? i think panic is more appropriate here than return

File: src/workers/postgres.rs [hunk 2]
```diff
@@ -23,36 +23,73 @@ use crate::encoding::BinaryEncoding;
 
 lazy_static::lazy_static! {
     pub static ref MESSAGES_IN_POSTGRES_CHANNEL: GenericGauge<prometheus::core::AtomicI64> = register_int_gauge!(opts!("literpc_messages_in_postgres", "Number of messages in postgres")).unwrap();
+    pub static ref POSTGRES_SESSION_ERRORS: GenericGauge<prometheus::core::AtomicI64> = register_int_gauge!(opts!("literpc_session_errors", "Number of failures while establishing postgres session")).unwrap();
+}
+
+const MAX_QUERY_SIZE: usize = 200_000; // 0.2 mb
+
+trait SchemaSize {
+    const DEFAULT_SIZE: usize = 0;
```
Comment (mschneider) [inline src/workers/postgres.rs:32]: can use 1 as a default

File: src/workers/postgres.rs [hunk 3]
```diff
@@ -23,36 +23,73 @@ use crate::encoding::BinaryEncoding;
...
+    const MAX_SIZE: usize = 0;
+}
+
+const fn get_max_safe_inserts<T: SchemaSize>() -> usize {
+    let Some(n) = MAX_QUERY_SIZE.checked_div(T::DEFAULT_SIZE) else {
+        return 0;
```
Comment (mschneider) [inline src/workers/postgres.rs:38]: here as well

File: src/workers/postgres.rs [hunk 4]
```diff
@@ -23,36 +23,75 @@ use crate::encoding::BinaryEncoding;
...
 pub struct PostgresTx {
-    pub signature: String,
-    pub recent_slot: i64,
-    pub forwarded_slot: i64,
-    pub forwarded_local_time: DateTime<Utc>,
+    pub signature: String,                   // 88 bytes
+    pub recent_slot: i64,                    // 8 bytes
+    pub forwarded_slot: i64,                 // 8 bytes
+    pub forwarded_local_time: DateTime<Utc>, // 8 bytes
     pub processed_slot: Option<i64>,
```
Comment (godmodegalactus) [inline src/workers/postgres.rs:58]: Option should be 1 bytes or 9 bytes right?

File: src/workers/postgres.rs [hunk 5]
```diff
@@ -327,37 +388,59 @@ impl Postgres {
                     }
                 }
 
-                let Ok(session) = self.get_session().await else {
+                if tx_batch.is_empty() && block_batch.is_empty() && update_batch.is_empty() {
+                    tokio::time::sleep(Duration::from_millis(400)).await;
```
Comment (godmodegalactus) [inline src/workers/postgres.rs:392]: add continue over here.

File: src/workers/postgres.rs [hunk 6]
```diff
@@ -299,23 +343,40 @@ impl Postgres {
...
+
+            let mut tx_batch: Vec<PostgresTx> = Vec::with_capacity(TX_MAX_CAPACITY);
+            let mut block_batch: Vec<PostgresBlock> = Vec::with_capacity(BLOCK_MAX_CAPACITY);
+            let mut update_batch =
+                Vec::<(PostgresUpdateTx, String)>::with_capacity(UPDATE_MAX_CAPACITY);
+
+            let mut session_establish_error = false;
 
             loop {
                 loop {
```
Comment (godmodegalactus) [inline src/workers/postgres.rs:358]: Could move all the vectors in the loop so that they are cleared at the end of the loop.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: src/workers/postgres.rs (hunk 4 above)
Comment (aniketfuryrocks) [reply src/workers/postgres.rs:58]: It is not sent in the query. I am not calculating struct size, but the size of the actual query

File: src/workers/postgres.rs (hunk 5 above)
Comment (aniketfuryrocks) [reply src/workers/postgres.rs:392]: oh yeah good catch

File: src/workers/postgres.rs (hunk 6 above)
Comment (aniketfuryrocks) [reply src/workers/postgres.rs:358]: no we don't want reallocation of memory. After a fixed capacity is allocated, just clear it and re use it. More efficient.

File: src/workers/postgres.rs (hunk 2 above)
Comment (aniketfuryrocks) [reply src/workers/postgres.rs]: done

File: bench/src/metrics.rs (hunk 1 above)
Comment (aniketfuryrocks) [reply bench/src/metrics.rs:40]: avg metric. if there were no runs then bench exits with a log

File: src/workers/postgres.rs [hunk 7]
```diff
@@ -327,37 +388,60 @@ impl Postgres {
                     }
                 }
 
-                let Ok(session) = self.get_session().await else {
+                if tx_batch.is_empty() && block_batch.is_empty() && update_batch.is_empty() {
+                    tokio::time::sleep(Duration::from_millis(400)).await;
```
Comment (mschneider) [inline src/workers/postgres.rs:392]: lower sleep here to 50ms, no need to wait so long

Comment (aniketfuryrocks) [reply src/workers/postgres.rs]: fixed

**Pull request merged**